lib/validator.js

```diff
@@ -0,0 +1,159 @@
+    for (const def of all) {
+      if (def.partial) {
+        const array = partials.get(def.name);
+        if (array) {
+          array.push(def);
+        } else {
+          partials.set(def.name, [def]);
+        }
+        continue;
+      }
+      if (!def.name) {
+        continue;
+      }
+      if (!unique.has(def.name)) {
+        unique.set(def.name, def);
+      } else {
+        duplications.add(def);
+      }
+    }
+    return { all, unique, partials, duplications };
+  }
+
+  function* checkDuplicatedNames({ unique, duplications }) {
+    for (const dup of duplications) {
+      const { name } = dup;
+      const message = `The name "${name}" of type "${unique.get(name).type}" was already seen`;
+      yield error(dup.source, dup.tokens.name, dup, message);
+    }
+  }
+
+  function* checkInterfaceMemberDuplication(defs) {
+    const interfaces = [...defs.unique.values()].filter(def => def.type === "interface");
+    const includesMap = getIncludesMap();
+
+    for (const i of interfaces) {
+      yield* forEachInterface(i);
+    }
+
+    function* forEachInterface(i) {
+      const opNames = getOperationNames(i);
+      const partials = defs.partials.get(i.name) || [];
+      const mixins = includesMap.get(i.name) || [];
+      for (const partial of partials) {
+        yield* forEachExtension(partial, opNames, i);
+      }
+      for (const mixin of mixins) {
+        yield* forEachExtension(mixin, opNames, i);
+      }
+    }
+
+    function* forEachExtension(ext, names, base) {
+      for (const op of ext.members.filter(mem => mem.type === "operation")) {
+        const name = (op.body && op.body.name) ? op.body.name.value : "";
+        if (name && names.has(name)) {
+          const message = `The operation "${name}" has already been defined in the base interface "${base.name}"`;
+          yield error(ext.source, op.body.tokens.name, ext, message);
+        }
+      }
+    }
+
+    function getOperationNames(i) {
+      const names = i.members
+        .filter(({type}) => type === "operation")
+        .map(op => (op.body && op.body.name) ? op.body.name.value : "");
+      return new Set(names);
+    }
+
+    function getIncludesMap() {
+      const map = new Map();
+      const includes = defs.all.filter(def => def.type === "includes");
+      for (const include of includes) {
+        const array = map.get(include.target);
+        const mixin = defs.unique.get(include.includes);
+        if (!mixin) {
+          continue;
+        }
+        if (array) {
+          array.push(mixin);
+        } else {
+          map.set(include.target, [mixin]);
+        }
+      }
+      return map;
+    }
+  }
+
+  function validate(ast) {
+    const defs = groupDefinitions(ast);
+    return [
+      ...checkDuplicatedNames(defs),
+      ...checkInterfaceMemberDuplication(defs)
+    ];
+  }
+
+  const obj = {
+    validate
+  };
+
+  if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
```

Comment:
I wonder if we should just ship the ES6 Module and not UMD this? I guess we’d still need to bundle it for distribution but might be nice to ship just ES6.

Reply:
I'd like to do that, will open a PR adding webpack.